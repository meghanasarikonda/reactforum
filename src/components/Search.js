import React, { Component } from 'react';
import { Button, Modal, ModalBody, ModalFooter } from 'reactstrap';
import Suggestions from './Suggestions';
import PropTypes from 'prop-types';

class Search extends Component {

  constructor(props) {
    super(props);

    this.state = {
      query: '',
      results: [],
    }

    this.toggle = this.toggle.bind(this);
    this.handleSuggestionClick = this.handleSuggestionClick.bind(this);
  }

  toggle() {
    this.setState(prevState => ({
      modal: !prevState.modal
    }))
  }

  async handleSuggestionClick(e) {
    
      e.preventDefault();
      e.stopPropagation();
      console.log(e.target.dataset.id, 'cool');
      this.props.handleUsernameClick(e);
      setTimeout(this.toggle, 1000)
      
    
    
  }

  getInfo() {
    console.log(this.props.usernameArr, 'usernameArr')
    let query = this.state.query;
    let results = [];

    for (var id = 1; id < this.props.usernameArr.length; id++) {
      let username = this.props.usernameArr[id];
        results.push({username, id});
    }

    let result = results.filter(({username}) => {
      if((username.slice(0, query.length)).toLowerCase() === query.toLowerCase()) {
        return username
      } 
      return false
    })
    console.log(result, 'resssss')
    this.setState({
      results: result
    })
  }
  

  handleInputChange = () => {
    this.setState({
      query: this.search.value
    }, () => {
      if (this.state.query && this.state.query.length > 1) {
        if (this.state.query.length % 2 === 0) {
          this.getInfo()
        }
      } 
    })
  }

  render() {
    return (
      <div>
        <Button
          onClick={this.toggle}
        >
          Search by Username
        </Button>
        <Modal isOpen={this.state.modal} toggle={this.toggle}>
          <ModalBody>
            <input 
              placeholder="Select or Search for..."
              ref={input => this.search = input}
              onChange={this.handleInputChange}
            />
            
            <Suggestions results={this.state.results} click={this.handleSuggestionClick} />
            
          </ModalBody>
          <ModalFooter>
            <Button color="secondary" onClick={() => this.setState({modal: false})}>Cancel</Button>
          </ModalFooter>
        
        </Modal>
      </div>
      
    )
  }
}

Search.propTypes = {
  usernameArr: PropTypes.array,
  toggleUser: PropTypes.func,
  handleUsernameClick: PropTypes.func 
}


export default Search;