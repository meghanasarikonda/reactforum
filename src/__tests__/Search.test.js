import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Search from '../components/Search';

function setup() {
  const myMock = jest.fn();
  myMock
    .mockReturnValueOnce('called once')
    .mockReturnValueOnce('called twice')
  const myMock2 = jest.fn();
  myMock2
    .mockReturnValueOnce('username called once')
    .mockReturnValueOnce('username called twice')

  const props = {
    toggleUser: myMock,
    handleUsernameClick: myMock2,
    usernameArr: []
  }

  const enzymeWrapper = shallow(<Search {...props} />)

  return {
    props,
    enzymeWrapper
  }
}

describe('Search', () => {

  it('should render correctly', () => {
    const { enzymeWrapper } = setup()
    expect(shallowToJson(enzymeWrapper)).toMatchSnapshot();
  });

  it('should have two children', () => {
    const { enzymeWrapper } = setup()
    expect(enzymeWrapper.children().length).toBe(2);
  });

  it('State should contain length 3', () => {
    const { enzymeWrapper } = setup()
    const instance = enzymeWrapper.instance();
    expect(Object.keys(instance.state).length).toBe(3);
  });

  it('should have modal', () => {
    const { enzymeWrapper } = setup()
    expect(enzymeWrapper.find('Modal').length).toBe(1);
  });

  it('should have modal footer', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find('ModalFooter').length).toBe(1)
  });

  it('Modal Footer should have a button', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find('ModalFooter').children().find('Button').length).toBe(1)
  });

  it('Modal should have isOpen property', () => {
    const { enzymeWrapper } = setup();
    enzymeWrapper.setState({ modal: true })
    expect(enzymeWrapper.find('Modal').props().isOpen).toBe(true)
  });

  it('Should have handleClick method', () => {
    const { enzymeWrapper } = setup();
    expect(Object.keys(enzymeWrapper.instance()).includes('handleClick')).toBe(true)
  });

  it('Should have handleInputChange method', () => {
    const { enzymeWrapper } = setup();
    expect(Object.keys(enzymeWrapper.instance()).includes('handleInputChange')).toBe(true)
  });

  it('Should have handleSuggestionClick method', () => {
    const { enzymeWrapper } = setup();
    expect(Object.keys(enzymeWrapper.instance()).includes('handleSuggestionClick')).toBe(true)
  });

  it('modal should open on handleClick event', () => {
    const { enzymeWrapper } = setup();
    enzymeWrapper.find('Button').at(0).simulate('click');
    expect(enzymeWrapper.state().modal).toBe(true)
  });

})

describe('Modal body', () => {

  it('should have input tag', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find('ModalBody').exists('input')).toBe(true);
  })

  it('should have onChange and onClick handlers', () => {
    const props = {
      toggleUser: jest.fn,
      handleUsernameClick: jest.fn,
      usernameArr: []
    }
  
    const enzymeWrapper = mount(<Search {...props} />)
    let search = {}
    search.value = 'Bret';
    enzymeWrapper.simulate('change')
    expect(!!(enzymeWrapper.state().query)).toBe(false);
    expect(typeof (enzymeWrapper.instance().getInfo)).toBe('function');
  })

  it('input text', () => {
    const { enzymeWrapper } = setup();
    console.log(enzymeWrapper.find('input').text())
  })

  it('should have suggestions as child', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find('ModalBody').exists('Suggestions')).toBe(true)
  })

})