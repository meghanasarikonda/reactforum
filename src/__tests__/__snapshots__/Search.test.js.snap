// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render correctly 1`] = `
<div>
  <Button
    color="secondary"
    onClick={[Function]}
    tag="button"
  >
    Search by Username
  </Button>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[Function]}
    zIndex={1050}
  >
    <ModalBody
      tag="div"
    >
      <input
        onChange={[Function]}
        placeholder="Select or Search for..."
      />
      <Suggestions
        click={[Function]}
        results={Array []}
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;
